astropy/utils/shapes.py

```diff
@@ -0,0 +1,329 @@
+
+        Parameters are as for :meth:`~numpy.ndarray.ravel`. Note that it is
+        not always possible to unravel an array without copying the data.
+        If you want an error to be raise if the data is copied, you should
+        should assign shape ``(-1,)`` to the shape attribute.
+        """
+        return self._apply('ravel', *args, **kwargs)
+
+    def flatten(self, *args, **kwargs):
+        """Return a copy with the array collapsed into one dimension.
+
+        Parameters are as for :meth:`~numpy.ndarray.flatten`.
+        """
+        return self._apply('flatten', *args, **kwargs)
+
+    def transpose(self, *args, **kwargs):
+        """Return an instance with the data transposed.
+
+        Parameters are as for :meth:`~numpy.ndarray.transpose`.  All internal
+        data are views of the data of the original.
+        """
+        return self._apply('transpose', *args, **kwargs)
+
+    @property
+    def T(self):
+        """Return an instance with the data transposed.
+
+        Parameters are as for :attr:`~numpy.ndarray.T`.  All internal
+        data are views of the data of the original.
+        """
+        if self.ndim < 2:
+            return self
+        else:
+            return self.transpose()
+
+    def swapaxes(self, *args, **kwargs):
+        """Return an instance with the given axes interchanged.
+
+        Parameters are as for :meth:`~numpy.ndarray.swapaxes`:
+        ``axis1, axis2``.  All internal data are views of the data of the
+        original.
+        """
+        return self._apply('swapaxes', *args, **kwargs)
+
+    def diagonal(self, *args, **kwargs):
+        """Return an instance with the specified diagonals.
+
+        Parameters are as for :meth:`~numpy.ndarray.diagonal`.  All internal
+        data are views of the data of the original.
+        """
+        return self._apply('diagonal', *args, **kwargs)
+
+    def squeeze(self, *args, **kwargs):
+        """Return an instance with single-dimensional shape entries removed
+
+        Parameters are as for :meth:`~numpy.ndarray.squeeze`.  All internal
+        data are views of the data of the original.
+        """
+        return self._apply('squeeze', *args, **kwargs)
+
+    def take(self, indices, axis=None, mode='raise'):
+        """Return a new instance formed from the elements at the given indices.
+
+        Parameters are as for :meth:`~numpy.ndarray.take`, except that,
+        obviously, no output array can be given.
+        """
+        return self._apply('take', indices, axis=axis, mode=mode)
+
+    # Functions that change shape or essentially do indexing.
+    _APPLICABLE_FUNCTIONS = {
+        np.moveaxis, np.rollaxis,
+        np.atleast_1d, np.atleast_2d, np.atleast_3d, np.expand_dims,
+        np.broadcast_to, np.flip, np.fliplr, np.flipud, np.rot90,
+        np.roll, np.delete,
+        }
+
+    # Functions that for now should just pass through since they
+    # (sort of) work.  TODO: create a proper implementation.
```

Comment:
For these (at least `np.where`) it isn't clear what a "proper implementation" should be.  Consider:
```
# tm, tm0 are Time objects
np.where.(tm > tm0, tm, 5)
```
An object array seems like the only option.

But if the second arg is a `Time` (`np.where(tm1 > tm0, tm1, tm2)`) and you want the output to be a single `Time` object (same for coordinates), I suppose you can use `info.new_like` and setitem functionality to make this work in the limited cases where it *can* work. You do get into a bit of API inconsistency, but maybe that's OK?

Reply:
Agreed it is not completely obvious, though precedence from the numpy side is that the alternatives for where are cast to a common type (see below). This argues that one should always get a `Time` out and your example should just raise an error. But partially because it is not obvious what the right behaviour is, I deferred it... (for `Time` itself, I wouldn't know what to do with one time that has a `location`, and one that has not; this is going to be a problem for `concatenate` as well).
```
In [2]: a = np.arange(5.)                                                       

In [3]: np.where(a>3, a, True)                                                  
Out[3]: array([1., 1., 1., 1., 4.])
```